packages/react/src/SelectPanel/SelectPanel.tsx

```diff
@@ -328,40 +340,106 @@ export function SelectPanel({
-
-      return {
-        ...item,
-        role: 'option',
-        selected: 'selected' in item && item.selected === undefined ? undefined : isItemSelected,
-        onAction: (itemFromAction, event) => {
-          item.onAction?.(itemFromAction, event)
-
-          if (event.defaultPrevented) {
-            return
-          }
+    return items
+      .map(item => {
+        const isItemSelected = isMultiSelectVariant(selected) ? doesItemsIncludeItem(selected, item) : selected === item
+
+        return {
+          ...item,
+          role: 'option',
+          selected: 'selected' in item && item.selected === undefined ? undefined : isItemSelected,
+          onAction: (itemFromAction, event) => {
+            item.onAction?.(itemFromAction, event)
+
+            if (event.defaultPrevented) {
+              return
+            }
 
-          if (isMultiSelectVariant(selected)) {
-            const otherSelectedItems = selected.filter(selectedItem => !areItemsEqual(selectedItem, item))
-            const newSelectedItems = doesItemsIncludeItem(selected, item)
-              ? otherSelectedItems
-              : [...otherSelectedItems, item]
+            if (isMultiSelectVariant(selected)) {
+              const otherSelectedItems = selected.filter(selectedItem => !areItemsEqual(selectedItem, item))
+              const newSelectedItems = doesItemsIncludeItem(selected, item)
+                ? otherSelectedItems
+                : [...otherSelectedItems, item]
 
-            const multiSelectOnChange = onSelectedChange as SelectPanelMultiSelection['onSelectedChange']
-            multiSelectOnChange(newSelectedItems)
-            return
-          }
+              const multiSelectOnChange = onSelectedChange as SelectPanelMultiSelection['onSelectedChange']
+              multiSelectOnChange(newSelectedItems)
+              return
+            }
 
-          // single select
-          const singleSelectOnChange = onSelectedChange as SelectPanelSingleSelection['onSelectedChange']
-          singleSelectOnChange(item === selected ? undefined : item)
-          onClose('selection')
-        },
-      } as ItemProps
-    })
-  }, [onClose, onSelectedChange, items, selected])
+            // single select
+            const singleSelectOnChange = onSelectedChange as SelectPanelSingleSelection['onSelectedChange']
+            singleSelectOnChange(item === selected ? undefined : item)
+            onClose('selection')
+          },
+        } as ItemProps
+      })
+      .sort((itemA, itemB) => {
+        // itemA is selected (for sorting purposes) if an object in selectedOnSort matches every property of itemA, except for the selected property
+        const itemASelected = selectedOnSort.some(item =>
+          Object.entries(item).every(([key, value]) => {
+            if (key === 'selected') {
+              return true
+            }
+            return itemA[key as keyof ItemProps] === value
+          }),
+        )
```

Comment:
this seems expensive :(

Reply:
Wanted to ask if it would be possible for the sort order to be maintained as a stateful value and compute it `onClose` or `onOpen` instead of having to run the sort every render?

It seemed like the logic that we have for this is something like:

- When open, keep the ordering of items constant
- When closing (or when re-opening), sort the items by:
  - selected items come first
    - sort selected items by group, if applicable
    - otherwise, sort by alphanumeric
  - otherwise, sort by alphanumeric

Does that track with what you're seeing?